File: ui/src/lib/api/training.ts
/**
 * Typed Training / Schulung client (P5).
 */

import { client } from "@/client/client.gen";

export type TrainingMode = "shadow" | "text" | "voice";
export type Difficulty = "beginner" | "intermediate" | "advanced";

export type TrainingModule = {
  id: number;
  organization_id: number;
  title: string;
  description: string;
  mode: TrainingMode | string;
  workflow_id: number | null;
  workflow_name: string;
  script_entry_id: number | null;
  success_codes: string[];
  tags: string[];
  difficulty: string;
  pass_score: number;
  content: Record<string, unknown>;
  is_published: boolean;
  created_by_user_id: number;
  created_at: string | null;
  best_score: number | null;
  attempts_count: number;
  completed: boolean;
};

export type TrainingModuleList = {
  total: number;
  items: TrainingModule[];
};

export type TrainingAttempt = {
  id: number;
  module_id: number;
  module_title: string;
  user_id: number;
  mode: string;
  score: number;
  passed: boolean;
  result: Record<string, unknown>;
  workflow_run_id: number | null;
  created_at: string | null;
};

export type TrainingProgress = {
  organization_id: number;
  user_id: number;
  modules_total: number;
  modules_completed: number;
  completion_pct: number;
  average_best_score: number | null;
  attempts_total: number;
  modules: Array<{
    module_id: number;
    title: string;
    mode: string;
    difficulty: string;
    pass_score: number;
    attempts_count: number;
    best_score: number | null;
    last_score: number | null;
    completed: boolean;
    last_attempt_at: string | null;
  }>;
};

export type ShadowQuizQuestion = {
  id: string;
  prompt: string;
  options: Array<{ id: string; label: string }>;
  explanation?: string;
};

function errMsg(err: unknown, fallback: string): string {
  if (typeof err === "object" && err && "detail" in err) {
    return String((err as { detail: unknown }).detail);
  }
  if (err instanceof Error) return err.message;
  return fallback;
}

export async function listTrainingModules(opts?: {
  mode?: string;
  published_only?: boolean;
}): Promise<TrainingModuleList> {
  const res = await client.get({
    url: "/api/v1/training/modules",
    query: {
      published_only: opts?.published_only ?? true,
      ...(opts?.mode ? { mode: opts.mode } : {}),
      limit: 100,
    },
  });
  if (res.error) throw new Error(errMsg(res.error, "Modules load failed"));
  return res.data as TrainingModuleList;
}

export async function createTrainingModule(body: {
  title: string;
  description?: string;
  mode: TrainingMode;
  workflow_id?: number | null;
  success_codes?: string[];
  tags?: string[];
  difficulty?: Difficulty;
  pass_score?: number;
  content?: Record<string, unknown>;
  is_published?: boolean;
}): Promise<TrainingModule> {
  const res = await client.post({
    url: "/api/v1/training/modules",
    body,
  });
  if (res.error) throw new Error(errMsg(res.error, "Create module failed"));
  return res.data as TrainingModule;
}

export async function fetchTrainingProgress(): Promise<TrainingProgress> {
  const res = await client.get({ url: "/api/v1/training/progress" });
  if (res.error) throw new Error(errMsg(res.error, "Progress failed"));
  return res.data as TrainingProgress;
}

export async function completeShadowModule(
  moduleId: number,
  answers: Array<{ question_id: string; selected_option_ids: string[] }>,
): Promise<TrainingAttempt> {
  const res = await client.post({
    url: `/api/v1/training/modules/${moduleId}/shadow/complete`,
    body: { answers },
  });
  if (res.error) throw new Error(errMsg(res.error, "Shadow complete failed"));
  return res.data as TrainingAttempt;
}

export async function runTextDrill(
  moduleId: number,
): Promise<TrainingAttempt> {
  const res = await client.post({
    url: `/api/v1/training/modules/${moduleId}/text/run`,
    body: {},
  });
  if (res.error) throw new Error(errMsg(res.error, "Text drill failed"));
  return res.data as TrainingAttempt;
}

export async function listMyAttempts(moduleId?: number): Promise<{
  total: number;
  items: TrainingAttempt[];
}> {
  const res = await client.get({
    url: "/api/v1/training/attempts",
    query: {
      limit: 50,
      ...(moduleId != null ? { module_id: moduleId } : {}),
    },
  });
  if (res.error) throw new Error(errMsg(res.error, "Attempts failed"));
  return res.data as { total: number; items: TrainingAttempt[] };
}

export type VoiceDrillStartResponse = {
  module_id: number;
  workflow_id: number;
  workflow_run_id: number;
  mode: string;
  max_duration_hint_seconds: number;
  signaling_path: string;
  guards: Record<string, unknown>;
  message: string;
};

export async function startVoiceDrill(
  moduleId: number,
  opts?: { max_duration_seconds?: number },
): Promise<VoiceDrillStartResponse> {
  const res = await client.post({
    url: `/api/v1/training/modules/${moduleId}/voice/start`,
    body: opts || {},
  });
  if (res.error) throw new Error(errMsg(res.error, "Voice start failed"));
  return res.data as VoiceDrillStartResponse;
}

export async function completeVoiceDrill(
  moduleId: number,
  workflowRunId: number,
  includeQa = true,
): Promise<TrainingAttempt> {
  const res = await client.post({
    url: `/api/v1/training/modules/${moduleId}/voice/complete`,
    body: { workflow_run_id: workflowRunId, include_qa: includeQa },
  });
  if (res.error) throw new Error(errMsg(res.error, "Voice complete failed"));
  return res.data as TrainingAttempt;
}

File: api/schemas/training.py
"""Training / Schulung MVP schemas (P5)."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


class ShadowQuizOption(BaseModel):
    id: str
    label: str


class ShadowQuizQuestion(BaseModel):
    id: str
    prompt: str
    options: list[ShadowQuizOption] = Field(default_factory=list)
    # correct option id(s) — not returned on learner GET (stripped in API)
    correct_option_ids: list[str] = Field(default_factory=list)
    explanation: str = ""


class ShadowContent(BaseModel):
    script_excerpt: str = ""
    learning_points: list[str] = Field(default_factory=list)
    quiz: list[ShadowQuizQuestion] = Field(default_factory=list)


class TextDrillTurn(BaseModel):
    user: str = Field(min_length=1)
    assertions: list[dict[str, Any]] = Field(default_factory=list)


class TextDrillContent(BaseModel):
    """Mirrors TextEvalScenario shape for reuse of the text harness."""

    scenario_name: str = "training-drill"
    initial_context: dict[str, Any] = Field(default_factory=dict)
    turns: list[TextDrillTurn] = Field(default_factory=list)
    final_assertions: list[dict[str, Any]] = Field(default_factory=list)


class TrainingModuleCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str = ""
    mode: Literal["shadow", "text", "voice"] = "shadow"
    workflow_id: Optional[int] = None
    script_entry_id: Optional[int] = None
    success_codes: list[str] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    difficulty: Literal["beginner", "intermediate", "advanced"] = "beginner"
    pass_score: float = Field(default=70.0, ge=0, le=100)
    content: dict[str, Any] = Field(default_factory=dict)
    is_published: bool = True


class TrainingModuleUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=255)
    description: Optional[str] = None
    mode: Optional[Literal["shadow", "text", "voice"]] = None
    workflow_id: Optional[int] = None
    script_entry_id: Optional[int] = None
    success_codes: Optional[list[str]] = None
    tags: Optional[list[str]] = None
    difficulty: Optional[Literal["beginner", "intermediate", "advanced"]] = None
    pass_score: Optional[float] = Field(default=None, ge=0, le=100)
    content: Optional[dict[str, Any]] = None
    is_published: Optional[bool] = None


class TrainingModuleResponse(BaseModel):
    id: int
    organization_id: int
    title: str
    description: str
    mode: str
    workflow_id: Optional[int] = None
    workflow_name: str = ""
    script_entry_id: Optional[int] = None
    success_codes: list[str] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    difficulty: str = "beginner"
    pass_score: float = 70.0
    content: dict[str, Any] = Field(default_factory=dict)
    is_published: bool = True
    created_by_user_id: int
    created_at: Optional[datetime] = None
    # learner progress (optional, filled when requested)
    best_score: Optional[float] = None
    attempts_count: int = 0
    completed: bool = False


class TrainingModuleListResponse(BaseModel):
    total: int
    items: list[TrainingModuleResponse]


class ShadowAnswer(BaseModel):
    question_id: str
    selected_option_ids: list[str] = Field(default_factory=list)


class ShadowCompleteRequest(BaseModel):
    answers: list[ShadowAnswer] = Field(default_factory=list)


class TextDrillRunRequest(BaseModel):
    """Optional overrides when running a text drill (defaults from module content)."""

    initial_context: Optional[dict[str, Any]] = None


class QuizItemResult(BaseModel):
    question_id: str
    correct: bool
    selected: list[str] = Field(default_factory=list)
    expected: list[str] = Field(default_factory=list)
    explanation: str = ""




class VoiceDrillStartRequest(BaseModel):
    """Optional overrides when starting a voice drill session."""

    max_duration_seconds: int = Field(default=90, ge=15, le=180)
    initial_context: Optional[dict[str, Any]] = None


class VoiceDrillStartResponse(BaseModel):
    module_id: int
    workflow_id: int
    workflow_run_id: int
    mode: str = "voice"
    max_duration_hint_seconds: int
    signaling_path: str
    guards: dict[str, Any] = Field(default_factory=dict)
    message: str = (
        "Connect via WebRTC, speak, hang up, then POST .../voice/complete."
    )


class VoiceDrillCompleteRequest(BaseModel):
    workflow_run_id: int
    include_qa: bool = True


class TrainingAttemptResponse(BaseModel):
    id: int
    module_id: int
    module_title: str = ""
    user_id: int
    mode: str
    score: float
    passed: bool
    result: dict[str, Any] = Field(default_factory=dict)
    workflow_run_id: Optional[int] = None
    created_at: Optional[datetime] = None


class TrainingAttemptListResponse(BaseModel):
    total: int
    items: list[TrainingAttemptResponse]


class ModuleProgressItem(BaseModel):
    module_id: int
    title: str
    mode: str
    difficulty: str
    pass_score: float
    attempts_count: int = 0
    best_score: Optional[float] = None
    last_score: Optional[float] = None
    completed: bool = False
    last_attempt_at: Optional[datetime] = None


class TrainingProgressResponse(BaseModel):
    organization_id: int
    user_id: int
    modules_total: int
    modules_completed: int
    completion_pct: float
    average_best_score: Optional[float] = None
    attempts_total: int
    modules: list[ModuleProgressItem] = Field(default_factory=list)
